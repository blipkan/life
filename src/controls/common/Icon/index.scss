.icon {
  display: flex;
  border-radius: 50%;
  width: 42px;
  height: 42px;
  justify-content: center;
  align-items: center;
  user-select: none;
  color: #000;
  margin: 16px;
  transition: all 0.3s ease-in-out;

  img {
    width: 20px;
    height: 20px;
  }
}
