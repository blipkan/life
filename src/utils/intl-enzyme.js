import React from 'react'
import { IntlProvider } from 'react-intl'
import flatten from 'flat'

const defaultLocale = 'ru'
const locale = defaultLocale

const messages = require(`../../src/messages/${locale}.json`)

export const withIntl = component => {
  return (
    <IntlProvider locale={locale} messages={flatten(messages)}>
      {React.cloneElement(component)}
    </IntlProvider>
  )
}
