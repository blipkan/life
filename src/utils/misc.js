import keys from 'lodash/keys'
import isObject from 'lodash/isObject'

export const makeId = (length = 16) =>
  parseInt(
    Math.ceil(Math.random() * Date.now())
      .toPrecision(length)
      .toString()
      .replace('.', '')
  )

export const uniqArray = (array = []) => [...new Set(array)]

export const compactArray = (array = []) => array.filter(e => e)

export const consoleMessage = (message, messageType) => {
  const consoleMethod = ['error', 'warn', 'log', 'info'].includes(messageType)
    ? messageType
    : 'error'
  console[consoleMethod](message)
}

export const checkKey = (obj, key, errorMessageType, errorMessage) => {
  if (!isObject(obj)) {
    const message =
      errorMessage || `${JSON.stringify(obj)} cannot have key '${key}'`
    consoleMessage(message, errorMessageType)
    return
  }

  let result
  try {
    result = obj[key]
    if (!result) {
      const message =
        errorMessage ||
        `invalid key '${key}' for object: ${JSON.stringify(
          obj
        )}. Available keys: ${keys(obj)}`
      consoleMessage(message, errorMessageType)
    }
  } catch (e) {
    consoleMessage(e.message, errorMessageType)
  }

  return result
}

export const getIntlText = (formatMessage, messages = {}, key) => {
  const message = messages[key]
  return message ? formatMessage(message) : key
}
