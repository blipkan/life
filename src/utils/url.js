/**
 * @description get array from url search parameter
 * @param {string} paramName  parameter name
 * @returns {Array<string>}
 */
export const getValueFromUrlSearch = paramName => {
  const url = new URL(location.href)
  return url.searchParams.get(paramName)
}
