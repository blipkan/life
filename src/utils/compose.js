import { compactArray, uniqArray } from './misc'

export const Delimiter = {
  CLASSNAME: '__',
  MODIFIER: '_',
  ID: '_'
}

const createArrayCompositor = (
  base,
  delimiter = Delimiter.CLASSNAME,
  whiteSpaceAsDelimiter
) => {
  const strings = Array.isArray(base) ? uniqArray(compactArray(base)) : [base]

  const roots = !whiteSpaceAsDelimiter
    ? strings
    : strings.reduce((acc, string) => {
        return acc.concat(...string.split(' '))
      }, [])

  return (...siffixes) => {
    const path = siffixes.join(delimiter)
    return roots.map(
      id => `${id || ''}${id && path ? delimiter : ''}${path || ''}`
    )
  }
}

const createCompositor = (base, delimiter = Delimiter.CLASSNAME) => {
  const arrayCompositor = createArrayCompositor(base, delimiter)
  return (...siffixes) => arrayCompositor(...siffixes).join(' ')
}

const createCompositors = (
  className,
  id,
  {
    classNameDelimiter = Delimiter.CLASSNAME,
    idDelimiter = Delimiter.ID,
    modifierDelimiter = Delimiter.MODIFIER
  } = {}
) => {
  const classNameArrayCompositor = createArrayCompositor(
    className,
    classNameDelimiter,
    true
  )
  const classNameCompositor = (...siffixes) =>
    classNameArrayCompositor(...siffixes).join(' ')

  const modifierCompositor = (modifier, ...siffixes) => {
    const paths = classNameArrayCompositor(...siffixes)
    return paths
      .map(
        path =>
          `${path || ''}${
            modifier && path ? modifierDelimiter : ''
          }${modifier || ''}`
      )
      .join(' ')
  }

  const classNameAndModifierCompositor = (modifier, ...siffixes) => {
    const classNameString = classNameCompositor(...siffixes)
    const modifierString = modifierCompositor(modifier, ...siffixes)
    return `${classNameString} ${modifierString}`
  }

  const idCompositor = createCompositor(id, idDelimiter)

  return {
    id: idCompositor,
    className: classNameCompositor,
    modifier: modifierCompositor,
    classNameAndModifier: classNameAndModifierCompositor
  }
}

export const createCompose = (props = {}, defaults = {}) => {
  const { className, domId } = props
  return createCompositors(
    [defaults.className, className],
    domId || defaults.domId
  )
}
