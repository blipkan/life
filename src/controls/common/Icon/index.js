import React from 'react'
import * as PropTypes from 'prop-types'

import { createCompose } from 'utils/compose'
import getIcon from 'images/icons'

import './index.scss'

export const Defaults = {
  className: 'icon',
  domId: 'icon',
  iconKey: 'blank'
}

const Icon = props => {
  const { iconKey = Defaults.iconKey, hint } = props

  const compose = createCompose(props, Defaults)

  const title = hint || iconKey

  return (
    <div className={compose.className()} id={compose.id()} title={title}>
      <img
        src={getIcon(iconKey)}
        title={title}
        alt={title}
        id={compose.id('img')}
      />
    </div>
  )
}

Icon.propTypes = {
  domId: PropTypes.string,
  className: PropTypes.string,
  iconKey: PropTypes.string,
  hint: PropTypes.string
}

export default Icon
