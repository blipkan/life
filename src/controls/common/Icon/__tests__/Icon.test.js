/* eslint-env jest */

import React from 'react'
import { mount, shallow, render } from 'enzyme'

import { createCompose } from 'utils/compose'

import Icon, { Defaults } from '../index'

describe('Icon', () => {
  const DOMID = 'DOMID'
  const CLASSNAME = 'CLASSNAME'
  const ICON_KEY = 'ICON_KEY'
  const HINT = 'HINT'
  const Properties = {
    className: CLASSNAME,
    domId: DOMID,
    iconKey: ICON_KEY,
    hint: HINT
  }

  it('render component', () => {
    const component = shallow(<Icon />)
    expect(component).toMatchSnapshot()
  })

  it('render props', () => {
    const props = { ...Properties }
    const compose = createCompose(props, Defaults)

    const component = shallow(<Icon {...props} />)
    const div = component.find('div')
    const img = component.find('img')

    expect(component).toMatchSnapshot()

    expect(div.prop('id')).toBe(compose.id())
    expect(div.prop('className')).toBe(compose.className())
    expect(div.prop('title')).toBe(props.hint)

    expect(img.prop('id')).toBe(compose.id('img'))
    expect(img.prop('alt')).toBe(props.hint)
    expect(img.prop('title')).toBe(props.hint)
  })

  it('handle empty hint (show iconKey)', () => {
    const props = { ...Properties, hint: undefined }
    const component = shallow(<Icon {...props} />)
    const div = component.find('div')
    const img = component.find('img')

    const expectedHint = props.iconKey
    expect(div.prop('title')).toBe(expectedHint)
    expect(img.prop('alt')).toBe(expectedHint)
    expect(img.prop('title')).toBe(expectedHint)
  })

  it('handle empty iconKey (use default=blank)', () => {
    const props = { ...Properties, iconKey: undefined, hint: null }
    const component = shallow(<Icon {...props} />)

    const div = component.find('div')
    const img = component.find('img')

    const expectedHint = Defaults.iconKey
    expect(div.prop('title')).toBe(expectedHint)
    expect(img.prop('alt')).toBe(expectedHint)
    expect(img.prop('title')).toBe(expectedHint)
    expect(typeof img.prop('src')).toBe('string')
  })
})
