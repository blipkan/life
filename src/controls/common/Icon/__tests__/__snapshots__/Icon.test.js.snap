// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon render component 1`] = `
<div
  className="icon"
  id="icon"
  title="blank"
>
  <img
    alt="blank"
    id="icon_img"
    src="test-file-stub"
    title="blank"
  />
</div>
`;

exports[`Icon render props 1`] = `
<div
  className="icon CLASSNAME"
  id="DOMID"
  title="HINT"
>
  <img
    alt="HINT"
    id="DOMID_img"
    src="test-file-stub"
    title="HINT"
  />
</div>
`;
